thrasher-corp/gocryptotrader
# Config: Add versioning

# Config

* Add config versioning manager. Motivation:
  * Huobi would like to move futures to coinmargined futures, but we don't want to lose the user's existing config
  * We have a number of places we have deprecated config clutter, and pointers to native types, purely so we can upgrade it. 
* Version1 handles the pairs manager format migration
* Version2 handles GDAX to Coinbase rename
* Version3 will handle AssetEnabled, but keeping this initial PR minimal
* Upgrade configtest to Version2
* Not upgrading config_example because it would jump over some null changes - I'll handle those in Version4 quickly
* Add -decrypt and -upgrade to cmd/config, no longer explicitly decrypts, or automatically reverses what was there
* Add -edit to edit a file in place
* -in instead of -infile and -out instead of outfile
* Do not echo input to user when entering encryption key
* 🐛 Fix encryption keys not allowing spaces
* This has some light touch improvements to encryption handling. I think I'm going to have to circle back to that separately later.

**Note Well:** The test failure in `TestConfigAllJsonResponse` is fixed by the dependency below

# Dependencies
* #1593 fixes `TestConfigAllJsonResponse`

## Type of change
